### PR TITLE
fix: harden signature reconstruction

### LINKED ISSUE #933: feat: post-reconstruction signature verification with fallback (PR #890 split 3/4)
## Context

Part 2 of the [PR #890 split](https://github.com/sigp/anchor/issues/931), per [review comment](https://github.com/sigp/anchor/pull/890#issuecomment-4118041676). Depends on #932.

## Scope

Core spec compliance (~250 LOC prod + ~400 LOC tests, 3 crates). Implements `ReconstructSignature` + `FallBackAndVerifyEachSignature`.

### Files

| File | Change |
|------|--------|
| `client/src/lib.rs` | Pass `database.clone()` to signature collector (1 line) |
| `validator_store/src/lib.rs` | Add `validator_pubkey` to `ValidatorSigningData` (1 line) |
| `signature_collector/Cargo.toml` | Add `metrics` + `database` deps |
| `signature_collector/src/lib.rs` | `try_combine_and_verify`, `verify_reconstructed_signature`, `find_invalid_shares`, `fetch_share_pubkeys`; `CombineOutcome` enum; `while` loop retry; DB wiring through manager; `SignatureCollecting` trait |
| `signature_collector/src/metrics.rs` | `SIGNATURE_VERIFICATION_FAILURES_TOTAL` counter |
| `signature_collector/src/tests.rs` | Unit tests for `find_invalid_shares`, `try_combine_and_verify`; integration tests for happy path + fallback eviction/recovery |

### Key design decisions
- `CombineOutcome` enum for explicit control flow (Success / CombineFailed / VerificationFailed)
- `while` loop replaces `if` to retry after evicting invalid shares
- `SignatureCollecting` trait for testability
- `spawn_blocking` for DB access to avoid blocking async runtime

### Known issue to address
- The `while` loop + eviction interaction when remaining shares still >= threshold needs review (flagged in original PR comment)

## Parent issue
- #931

### PR DESCRIPTION
**Pull request by shane-moore**
## Problem, Evidence, and Context

- Signature collection should only complete after the reconstructed signature matches the validator public key.
- Failed reconstruction attempts should remain recoverable when a valid quorum remains or later arrives.
- Closes #933.

## Change Overview

- Verify each reconstructed signature before caching it or notifying callers.
- After a reconstruction failure, load the validator share public keys through a bounded blocking path, remove shares that fail individual verification, and either retry once or wait for later shares.
- Keep the healthy path reconstruct-first, with no database lookup or per-share verification.
- Leave duplicate handling, wire formats, database schema, and notifier errors unchanged.

## Risks, Trade-offs, and Mitigations

- The fallback adds database and BLS work only after reconstruction fails. A manager-wide permit bounds concurrent blocking lookups.
- Inconsistent key data and lookup failures fail closed without caching or notifying.
- Focused tests cover verified completion, recovery after invalid shares, fatal inconsistencies, caching, and late callers.

## Validation

- `cargo test -p signature_collector -p anchor_validator_store --release --locked`
- `make cargo-fmt-check`
- `make lint`
- `git diff --check`

## Rollback

- Revert the commit. There are no schema, configuration, or data migrations.

### CODE REVIEW

**Comment by shane-moore** [comment]
@claude, please take a look

File: anchor/signature_collector/src/lib.rs [hunk 1]
```diff
@@ -573,88 +587,268 @@ impl<S: SlotClock + Clone + 'static> SignatureCollecting for Arc<SignatureCollec
...
+
+    warn!(
+        failure_kind,
+        ?combination_error,
+        share_count = state.signature_share.len(),
+        threshold,
+        "Reconstructed signature failed, verifying buffered shares"
+    );
+
+    let Some(share_pubkeys) = share_pubkey_loader.fetch(validator_pubkey).await else {
```
[the remark this answers is not in the document]
Comment (shane-moore) [reply anchor/signature_collector/src/lib.rs:687]: @claude, good observation. I traced the repeated-fallback path, including repeated validator entries in a batch, and agree that it can cause more than one lookup.

I am leaving this unchanged in this PR for three reasons:

1. Caching `PublicKeyBytes` avoids only the SQLite lookup and `spawn_blocking` handoff. `remove_invalid_shares` would still decompress keys and BLS-verify the full buffered set on every fallback.
2. The proposed cache has no invalidation key. A validator can be removed and re-added with new share keys while a collector is alive, and collector registration carries no cluster or configuration generation. An unversioned map could therefore reject current shares or retain an old configuration.
3. It does not help cold fan-out across collectors, which is the case the manager-wide semaphore is intended to bound.

The new fallback metric and aggregate logs give us evidence if repeated lookup cost becomes material. If it does, I would prefer a cache tied to a network-state generation, or coalescing duplicate entries alongside #934, rather than storing an unversioned database snapshot in `SignatureCollectorState`.

**Comment by shane-moore** [comment]
@claude, thanks for the thorough review. I am not making code changes for the two non-blocking observations.

The manager-wide `Semaphore(1)` is intentional because `NetworkDatabase` also has a one-connection pool. A slow lookup serializes database access either way. Acquiring the permit before `spawn_blocking` makes later fallbacks wait asynchronously instead of occupying Tokio blocking workers while they wait for that connection. If the pool size changes, or fallback telemetry shows sustained contention, the permit count can be revisited.

I replied inline on caching. In short, an unversioned share-key cache introduces invalidation semantics and removes only the database handoff, not repeated BLS verification. I would rather measure the new fallback telemetry first and use a versioned cache or duplicate coalescing if the cost proves material.

**Review by jnhsigmap — APPROVED** [review]
well considered and good a companion to upcoming randao work. it was good to learn about signature collector service in the process. I mainly focused on performance and test cases and outstanding edge cases and couldn't identify any probable weaknesses.
